### ISSUE TITLE
[RFC] Promote the Forward Security for KBS Protocol

### ISSUE THREAD

**Issue by Xynnn007** [confidential-containers/kbs#143]
#### Background

[KBS Attestation Protocol](https://github.com/confidential-containers/kbs/blob/main/docs/kbs_attestation_protocol.md) is a protocol proposed by CoCo Community. It is really an important protocol in CoCo community as it has the following functionalities:
1. Authenticates a KBS-Client based on remote-attestation evidences
2. Authorize the KBS-Client to access the confidential resources inside the KBS

Some of the security considerations of the protocol:
1. A server-side challenge prevents the client to reply the evidence
2. A tee key pair is used to encrypt the resource bytes to ensure the confidentiality
3. HTTPS is used to ensure the authtity of the KBS

The simplified RCAR interaction is as the following (C for client and S for KBS Server):
```
1. C-->S     Request := {}

2. C<--S     Challenge := { Nonce }

3. C-->S     Attestation := { Ev(Nonce, pubkey), pubkey }

4. C<--S     Response := { Token(pubkey, Ev) }
```

When it comes to resource retrievement, the interaction:
```
C-->S    request for resource `A`
C<--S    Enc(tee pubkey, symmetric key), Enc(symmetric key, data of `A`) 
```
Then the symmetric key can be decrypted by the tee privatekey of C, and data of `A` can be decrypted by the symmetric key.

Now the tee pubkey used to encrypt the JWE encryption key is bond to a Token. KBS certifies the token. However once the tee privatekey is leaked to an attacker, the attacker can decrypt all the JWE messages during the lifetime of the teepubkey, which violates [forward security](https://en.wikipedia.org/wiki/Forward_secrecy).

This problem can be more severe when a long time token is used.

#### Workaround

A fix-up is abondon JWE scheme. Instead, Diffie-Hellman key exchange should be introduced. When it comes to resource retrievement, the interaction should be
```
C-->S    Diffie-Hellman key exchange with all messages signed by tee private key. The S should ensure the authtity of the signer and integrity of the key share used to negotiate. Finally C and S achieve a common secret and derive it into a 
ephemeral key to encrypt the data.
C<--S    Enc(ephemeral key, data)
```

In this way, even if the signing key is leaked, the key shares captured by the attacker are not enough to derive the historical ephemeral keys, thus protecting the forward security.

We do not need to do this on application layer (s.t. over https). Instead, the step 4 in RCAR can be
```
4. C<--S     Response := { Token(cert, Ev), cert(pubkey) }
```
where the cert is a tls certificate signed by KBS. Then, the resource requests does not need to depends on a http session or a BearAuth header. Only need the KBS (web server) to open mTLS on resource access related APIs where the other side's pubkey should be the one it signs the cert.

cc @sameo @fitzthum @jialez0 @mkulke 

Can you share some point of views?

**Comment by fitzthum** [comment]
Yeah, basically we should add an ephemeral key to the KBS protocol rather than encrypting to the public key of the TEE.

**Comment by Xynnn007** [comment]
Some views about this.

We have some candidates
1. Use HPKE. This helps to protect KBS side PFS, but not the TEE side. Because once the TEE private key is exposed, attacker could use historical ephk and the tee pri key to generate the shared secret. We need ephemeral ECDHE to achieve PFS.
2. Use mTLS1.3. This is the standard way to achieve PFS.
3. Use Noise over HTTP. This can be a message level to achieve PFS.

**Comment by Xynnn007** [comment]
Let's consider two issues, and think about the correlation and independence of these two issues.
1. One is the client's authentication of the KBS identity, 
2. and the other is ensuring the forward security of the protocol.

To achieve 1, the only way is to provision some material for the client. For CoCo, initdata is the tunnel, and the material is KBS cert. This means that we have TLS.

To achieve 2, there are also two ways
- implement this at transport layer level, s.t. using TLS
- implement this at application layer level, s.t. promote current JWE to a PFS version. Options are using noise, or a customized JWE (risks/challenges for "inventing our own security") or so.

We can see the table

| Method | Server identity authentication | PFS |
|---|---|---|
|  transport layer level (mTLS) | Y | Y |
| application layer level | N | Y |


Therefore, there are two approaches to solving this problem.

1. Resolve them at the same time, s.t. KBS public key is always required. In this way, it's doable via using mTLS.
2. Solving two problems vertically. That means we should implement this at application level.

btw, I once make a demo with approach 1 and works fine.
